### PR TITLE
Next/previous background keyboard shortcuts

### PR DESCRIPTION
**Pull request by vanreece**
Binds , to previous background in list, and . to next background in
list. The thinking to my mind is that those are under the < and >
characters on a US keyboard, but that may not be international enough.

Speaking of which, I put those keys in core.yaml (en > background >
(next|previous)_background > key). Not really sure if that's the best
place for it or not, open to suggestions. Also, do we commit the built
json file "dist/locales/en.json" or not?

This takes the current list of backgrounds, in displayed order and
allows you to iterate through it forwards or backwards. It will only do
the custom background if there's a custom background defined.

### CODE REVIEW

File: modules/ui/background.js [hunk 1]
```diff
@@ -318,6 +316,45 @@ export function uiBackground(context) {
...
+        }
+
+        var nextBackgroundIndex = (foundIndex + offset + backgrounds.length) % backgrounds.length;
+        var nextBackground = backgrounds[nextBackgroundIndex];
+        if (nextBackground.id === 'custom' && !nextBackground.template()) {
+            nextBackgroundIndex = (nextBackgroundIndex + offset + backgrounds.length) % backgrounds.length;
+            nextBackground = backgrounds[nextBackgroundIndex];
+        }
+        chooseBackground(nextBackground);
+    }
```
Comment (Bonkles) [inline modules/ui/background.js:348]: This function uses `var` and `let`. There are plenty of places where 'var' is still in the iD code, but any new code should be as ES6-ified as possible. Replace `var` with `const` or `let` as appropriate!

File: modules/ui/background.js [hunk 2]
```diff
@@ -318,6 +316,45 @@ export function uiBackground(context) {
...
+        var backgrounds = getBackgrounds(function(d) { return !d.isHidden() && !d.overlay; });
+        backgrounds.sort(sortSources);
+        var currentBackground = context.background().baseLayerSource();
+        var foundIndex = -1;
+        for (let index = 0; index < backgrounds.length; index++) {
+            if (backgrounds[index] === currentBackground) {
+                foundIndex = index;
+                break;
+            }
+        }
```
Comment (Bonkles) [inline modules/ui/background.js:335]: This entire block can be replaced with a call of the array helper method `indexOf()` - it returns -1 if not found: 

`        const foundIndex = backgrounds.indexOf(currentBackground);`

File: data/core.yaml [hunk 3]
```diff
@@ -735,6 +735,10 @@ en:
       tooltip: Show advanced background information.
     fix_misalignment: Adjust imagery offset
     offset: "Drag anywhere in the gray area below to adjust the imagery offset, or enter the offset values in meters."
+    next_background:
```
[the remark this answers is not in the document]
Comment (bhousel) [reply data/core.yaml:738]: whoa cool - I never noticed the RapiD badges here..

File: modules/ui/background.js (hunk 1 above)
Comment (vanreece) [reply modules/ui/background.js:348]: image

File: data/core.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (vanreece) [reply data/core.yaml:738]: Oh yeah! shortcuts.json. I knew I'd forget something.

File: modules/ui/background.js (hunk 2 above)
Comment (vanreece) [reply modules/ui/background.js:335]: Sweet javascripty goodness. Thanks for that. :)

**Comment by vanreece** [comment]
@Bonkles Okay, incorporated feedback. Looking pretty good to me. I added the RapiD badge to the shortcuts, I guess I could try to land a similar change in iD and then eventually we could remove that badge if all goes well there, eh?

**Comment by Bonkles** [comment]
> @Bonkles Okay, incorporated feedback. Looking pretty good to me. I added the RapiD badge to the shortcuts, I guess I could try to land a similar change in iD and then eventually we could remove that badge if all goes well there, eh?

Yup, that sounds like a good way to play it! Thanks for addressing feedback too!

**Comment by Bonkles** [comment]
@vanreece what are your next steps with this PR? Get it merged in, test with internal testers in an internal build, then iterate? or is there more work you'd like to do on it?

File: modules/ui/background.js (hunk 1 above)
Comment (Bonkles) [reply modules/ui/background.js]: yasssssss

File: modules/ui/background.js [hunk 4]
```diff
@@ -318,6 +316,39 @@ export function uiBackground(context) {
...
+        return context.background()
+            .sources(context.map().extent(), context.map().zoom(), true)
+            .filter(filter);
+    }
+
+    function chooseBackgroundAtOffset(offset) {
+        const backgrounds = getBackgrounds(function(d) { return !d.isHidden() && !d.overlay; });
+        backgrounds.sort(sortSources);
+        const currentBackground = context.background().baseLayerSource();
+        const foundIndex = backgrounds.indexOf(currentBackground);
```
Comment (Bonkles) [inline modules/ui/background.js:329]: So terse. So *clean*. It makes my heart sing.

**Comment by vanreece** [comment]
@Bonkles okay, I cherry picked the two fixes that @bhousel had on the esri branch, which fixes the tests on master and here into this commit. How do I kick the CI

File: css/80_app_fb.css [hunk 5]
```diff
@@ -472,4 +472,31 @@ g.layer-ai-features.hide-rapid-buildings path.building  {
 
 .modal-section.rapid-checkbox.poweruser:hover div#tag-sources { 
     background-image: url('img/tag-sources.gif'); 
+}
+
+ul.layer-list.layer-background-list li button.background-favorite-button{
```
Comment (Bonkles) [inline css/80_app_fb.css:477]: format nit- add a space between the .css rule specifiers and the opening brace ⬇️ 

`ul.layer-list.layer-background-list li button.background-favorite-button {`

File: data/core.yaml [hunk 6]
```diff
@@ -2110,6 +2114,8 @@ en:
         title: "Display options"
         background: "Toggle background imagery pane"
         background_switch: "Switch back to last background"
+        background_previous: "Switch to previous background in list"
```
Comment (Bonkles) [inline data/core.yaml:2117]: Consider removing 'in list' from this description- I think 'switch to previous background' is perfectly adequate

File: modules/ui/background.js [hunk 7]
```diff
@@ -318,6 +354,39 @@ export function uiBackground(context) {
         }
     }
 
+    function getBackgrounds(filter) {
+        return context.background()
+            .sources(context.map().extent(), context.map().zoom(), true)
+            .filter(filter);
+    }
+
```
Comment (Bonkles) [inline modules/ui/background.js:362]: Yet another formatting nit- functions in iD have two spaces between them. NOT a big deal, but something to note!

**Review by Bonkles — APPROVED** [review]

**Pull request merged**